tools/amd_build/pyHIPIFY/hipify-python.py

```diff
@@ -0,0 +1,946 @@
+
+
+def walk_over_directory(rootpath, extensions, show_detailed=False, include_dirs=None):
+    """
+    Recursively walk over directory and call preprocessor on selected files.
+
+    Arguments)
+        extensions - A plist of file extensions ['cu', 'cuh', ..]
+
+        include_dirs - Directories under the rootpath that should be included in the walk.
+
+        show_detailed - Show a detailed summary of the transpilation process.
+    """
+
+
+    # Default argument for excluded directories.
+    if include_dirs is None:
+        include_dirs = []
+
+    # Compute the total number of files to be traversed.
+    total_files = 0
+    for (dirpath, _dirnames, filenames) in os.walk(rootpath):
+        if inside_included_directories(dirpath, rootpath, include_dirs):
+            for filename in filenames:
+                total_files += filename_ends_with_extension(filename, extensions)
+
+    current_file = 0
+
+    # Preprocessing statistics.
+    stats = {"unsupported_calls": [], "kernel_launches": []}
+
+    # Begin traversing the files.
+    for (dirpath, _dirnames, filenames) in os.walk(rootpath, topdown=True):
+        # Check if file ends with a valid extensions
+        if not inside_included_directories(dirpath, rootpath, include_dirs):
+            continue
+
+        for filename in filenames:
+            if filename_ends_with_extension(filename, extensions):
+                # Construct the file's full path
+                filepath = os.sep.join([dirpath, filename])
+
+                # Execute the preprocessor on the specified file.
+                preprocessor(filepath, stats)
+
+                # Update the progress
+                print(os.path.join(dirpath, filename))
+                update_progress_bar(total_files, current_file)
+
+                current_file += 1
+
+    print(bcolors.OKGREEN + "Successfully preprocessed all matching files." + bcolors.ENDC)
+
+    # Show detailed summary
+    if show_detailed:
+        compute_stats(stats, show_detailed)
+
+
+def compute_stats(stats):
+    unsupported_calls = set(cuda_call for (cuda_call, _filepath) in stats["unsupported_calls"])
+
+    # Print the number of unsupported calls
+    print("Total number of unsupported CUDA function calls: %d" % (len(unsupported_calls)))
+
+    # Print the list of unsupported calls
+    print(", ".join(unsupported_calls))
+
+    # Print the number of kernel launches
+    print("\nTotal number of replaced kernel launches: %d" % (len(stats["kernel_launches"])))
+
+def processKernelLaunches(string, stats):
+    """ Replace the CUDA style Kernel launches with the HIP style kernel launches."""
+    # Concat the namespace with the kernel names. (Find cleaner way of doing this later).
+    string = re.sub(r'([ ]+)(detail+)::[ ]+\\\n[ ]+', lambda inp: "%s%s::" % (inp.group(1), inp.group(2)), string)
```

Comment:
You don't need braces for a character class if there is only one character class.

Reply:
Surely you don't want `detail+` here; that will match `detaildetaildetail` which cannot be correct. `detail?` is probably more correct.

Reply:
done